File: 0099-recover-binary-search-tree/0099-recover-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    TreeNode first = null, second = null, prev = null;

    public void recoverTree(TreeNode root) {
        TreeNode cur = root;

        while (cur != null) {
            if (cur.left == null) {
                checkSwap(cur);
                cur = cur.right;
            } else {
                TreeNode predecessor = cur.left;
                while (predecessor.right != null && predecessor.right != cur) {
                    predecessor = predecessor.right;
                }

                if (predecessor.right == null) {
                    predecessor.right = cur;
                    cur = cur.left;
                } else {
                    predecessor.right = null;
                    checkSwap(cur);
                    cur = cur.right;
                }
            }
        }
        
        int tmp = first.val;
        first.val = second.val;
        second.val = tmp;
    }

    private void checkSwap(TreeNode cur) {
        if (prev != null && prev.val > cur.val) {
            if (first == null) {
                first = prev;
            }

            second = cur;
        }

        prev = cur;
    }
}

File: 0099-recover-binary-search-tree/0099-recover-binary-search-tree.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    fun recoverTree(root: TreeNode?): Unit {
        var first: TreeNode? = null
        var second: TreeNode? = null
        var prev: TreeNode? = null
        var cur = root

        while (cur != null) {
            if (cur.left == null) {
                if (prev != null && prev.`val` > cur.`val`) {
                    if (first == null) {
                        first = prev
                    }

                    second = cur
                }

                prev = cur
                cur = cur.right
            } else {
                var predecessor = cur.left
                while (predecessor?.right != null && predecessor.right != cur) {
                    predecessor = predecessor.right
                }

                if (predecessor?.right == null) {
                    predecessor?.right = cur
                    cur = cur.left
                } else {
                    predecessor.right = null
                    if (prev != null && prev.`val` > cur.`val`) {
                        if (first == null) {
                            first = prev
                        }

                        second = cur
                    }

                    prev = cur
                    cur = cur.right
                }
            }
        }

        if (first != null && second != null) {
            first.`val` = second.`val`.also { second.`val` = first.`val` }
        }
    }
}